### PR TITLE
Added Exponential Moving Average support to classification reference script

### PR DESCRIPTION
**Pull request by prabhat00155**
Resolves https://github.com/pytorch/vision/issues/4346.

cc @datumbox

### CODE REVIEW

File: references/classification/train.py
Comment (datumbox) [inline references/classification/train.py:324]: Is there a reason why you introduce an extra param instead of setting the default of `--model-ema-decay` to 0.0? In that case your EMA will be as if you run the normal validation and thus you could omit it.

File: references/classification/train.py
Comment (datumbox) [inline references/classification/train.py:208]: We might need to pass the `model_without_ddp` version of the model here.

**Review by datumbox — COMMENTED** [review]
Thanks @prabhat00155, overall it looks good. I left a couple of comments for your consideration. Let me know what you think.

Also do you plan to investigate how we can avoid loading two models on the GPU?

File: references/classification/utils.py
Comment (NicolasHug) [inline references/classification/utils.py:178]: I'm just wondering, what is the reason for having a `name` attribute, and why do we need it to be a `property`?
IDK if this is relevant but we can access an object's class name with `obj.__class__.__name__`

**Review by NicolasHug — COMMENTED** [review]
Thanks @prabhat00155 , I made some minor comments / questions

File: references/classification/utils.py
Comment (NicolasHug) [inline references/classification/utils.py:165]: Doc nits that don't really matter since we don't render the docs, but good to fix if we ever move this in the public API. Also good to know for future PRs:

- the first sentence should be in one line according to pep8. When this isn't respected, sphinx can create rendering issue (typically when displaying the summary of a function / class)
- since sphinx relies on rst format (not markdown like in github), the single backticks should be double backticks.
- I think there's a missing `_` at the end of the link below, although I haven't checked

File: references/classification/train.py
Comment (NicolasHug) [reply references/classification/train.py:324]: I don't have a strong opinion on this, but I cant think of 2 reasons to have both --model-ema-decay and  the --model-ema flag as currently done:

- the flag allows for --model-ema-decay to have a good default of 0.99 which hopefully should work fine for most models (that might actually not be true in practice though, IDK). Without this, users would have to come up with a value themselves.
- it may not be obvious for inexperienced users that `ema_decay == 0 => no EMA`. The flag helps disambiguate whether EMA is used at all

Again, no strong opinion. Both of these could actually be solved with a few sentences in the parameter description.

Comment (prabhat00155) [reply references/classification/train.py:324]: I guess @NicolasHug explained it well. Having `--model-ema` allows us to have a good default which users may choose to use. It also makes it simpler to use(not having to worry about choosing an ema_decay in your first run).

**Comment by prabhat00155** [comment]
> Thanks @prabhat00155 , I made some minor comments / questions
> 
> Also, it'd be good to validate the new code somehow. We don't really have tests for the references but perhaps it would be relevant to report the result of a model relying on the --model_avg option to compare to the baseline?

Thanks @NicolasHug! I ran tests locally(on CPU) on a toy dataset and on aws cluster for 5 epochs. I will run it to completion to verify the results before merging this PR.

**Comment by prabhat00155** [comment]
> Thanks @prabhat00155, overall it looks good. I left a couple of comments for your consideration. Let me know what you think.
> 
> Also do you plan to investigate how we can avoid loading two models on the GPU?

Thanks @datumbox! Yes, I was thinking of doing that in a follow-up PR.

**Comment by datumbox** [comment]
@prabhat00155 sounds good to me. 

The only thing worth addressing here is to pass the non parallel model in ema (see my earlier comment). I believe what you have now will fail to handle properly the checkpoints on a multi gpu setup. Worth confirming by doing two epochs on AWS and resuming from checkpoint. If it works you can leave as is.

Everything else are nits that can be done later on separate PRs.

File: references/classification/utils.py
Comment (prabhat00155) [reply references/classification/utils.py]: Yeah, I think it's unnecessary, thanks!

File: references/classification/utils.py
Comment (prabhat00155) [reply references/classification/utils.py]: Thanks, updated them.

**Review by datumbox — APPROVED** [review]
@prabhat00155 As we discussed offline, this is a great contribution so let's unblock the merge and push further investigations on follow up PRs.

File: references/classification/train.py
Comment (prabhat00155) [reply references/classification/train.py]: Updated it.

**Comment by prabhat00155** [comment]
> @prabhat00155 As we discussed offline, this is a great contribution so let's unblock the merge and push further investigations on follow up PRs.

Makes sense, thanks!

**Pull request merged**
